src/differential_expression/deseq2/script.R

```diff
@@ -0,0 +1,458 @@
+library(DESeq2)
+library(anndataR)
+library(hdf5r)
+
+## VIASH START
+par <- list(
+  input = paste0(
+    "resources_test/annotation_test_data/",
+    "TS_Blood_filtered_pseudobulk.h5mu"
+  ),
+  output_dir = "./test_deseq2_no_cellgroup/",
+  output_prefix = "deseq2_analysis",
+  input_layer = NULL,
+  modality = "rna",
+  obs_cell_group = "cell_type",
+  design_formula = "~ treatment",
+  contrast_column = "treatment",
+  contrast_values = c("ctrl", "stim"),
+  filter_genes_min_samples = NULL,
+  p_adj_threshold = 0.05,
+  log2fc_threshold = 0.0,
+  filter_gene_patterns = c(
+    "MIR\\d+", "AL\\d+", "LINC\\d+", "AC\\d+", "AP\\d+"
+  ),
+  var_gene_name = "feature_name"
+)
+meta <- list(resources_dir = "src/utils")
+## VIASH END
+
+cat("Starting DESeq2 analysis...\n")
+
+h5mu_to_h5ad <- function(h5mu_path, modality_name) {
+  tmp_path <- tempfile(fileext = ".h5ad")
+  mod_location <- paste("mod", modality_name, sep = "/")
+  h5src <- hdf5r::H5File$new(h5mu_path, "r")
+  h5dest <- hdf5r::H5File$new(tmp_path, "w")
+  # Copy over the child objects and the child attributes from root
+  # Root cannot be copied directly because it always exists and
+  # copying does not allow overwriting.
+  children <- hdf5r::list.objects(h5src,
+    path = mod_location,
+    full.names = FALSE, recursive = FALSE
+  )
+  for (child in children) {
+    h5dest$obj_copy_from(
+      h5src, paste(mod_location, child, sep = "/"),
+      paste0("/", child)
+    )
+  }
+  # Also copy the root attributes
+  root_attrs <- hdf5r::h5attr_names(x = h5src)
+  for (attr in root_attrs) {
+    h5a <- h5src$attr_open(attr_name = attr)
+    robj <- h5a$read()
+    h5dest$create_attr_by_name(
+      attr_name = attr,
+      obj_name = ".",
+      robj = robj,
+      space = h5a$get_space(),
+      dtype = h5a$get_type()
+    )
+  }
+  h5src$close()
+  h5dest$close()
+
+  tmp_path
+}
+
+# Check if expression data is normalized (row sums =~ 1)
+is_normalized <- function(layer) {
+  row_sums <- if (is(layer, "sparseMatrix") || is(layer, "dgCMatrix")) {
+    Matrix::rowSums(layer)
+  } else {
+    rowSums(layer)
+  }
+
+  all(abs(row_sums - 1) < 1e-6, na.rm = TRUE)
+}
+
+# Extract design factors from formula
+parse_design_formula <- function(design_formula) {
+  design_factors <- all.vars(as.formula(design_formula))
```

Comment:
What happens when a malformatted formula is provided? Maybe its something to check for and provide a proper error if required?